File: src/main/java/me/zeroeightsix/kami/util/ColourHolder.java
package me.zeroeightsix.kami.util;

import org.lwjgl.opengl.GL11;

import java.awt.*;

/**
 * Created by Gebruiker on 18/04/2017.
 */
public class ColourHolder {

    int r;
    int g;
    int b;
    int a;

    public ColourHolder(int r, int g, int b) {
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = 255;
    }

    public ColourHolder(int r, int g, int b, int a) {
        this.r = r;
        this.g = g;
        this.b = b;
        this.a = a;
    }

    public ColourHolder brighter() {
        return new ColourHolder(Math.min(r + 10, 255), Math.min(g + 10, 255), Math.min(b + 10, 255), getA());
    }

    public ColourHolder darker() {
        return new ColourHolder(Math.max(r - 10, 0), Math.max(g - 10, 0), Math.max(b - 10, 0), getA());
    }

    public void setGLColour() {
        setGLColour(-1, -1, -1, -1);
    }

    public void setGLColour(int dr, int dg, int db, int da) {
        GL11.glColor4f((float) (dr == -1 ? r : dr) / 255, (float) (dg == -1 ? g : dg) / 255, (float) (db == -1 ? b : db) / 255, (float) (da == -1 ? a : da) / 255);
    }

    public void becomeGLColour() {
//        FloatBuffer buffer = FloatBuffer.allocate(16);
//        GL11.glGetFloat(GL11.GL_CURRENT_COLOR, buffer);
//        becomeHex((int) buffer.get());
    }

    public void becomeHex(int hex) {
        setR((hex & 0xFF0000) >> 16);
        setG((hex & 0xFF00) >> 8);
        setB((hex & 0xFF));
        setA(255);
    }

    public static ColourHolder fromHex(int hex) {
        ColourHolder n = new ColourHolder(0, 0, 0);
        n.becomeHex(hex);
        return n;
    }

    public static int toHex(int r, int g, int b) {
        return (0xff << 24) | ((r & 0xff) << 16) | ((g & 0xff) << 8) | (b & 0xff);
    }

    public int toHex() {
        return toHex(r, g, b);
    }

    public int getB() {
        return b;
    }

    public int getG() {
        return g;
    }

    public int getR() {
        return r;
    }

    public int getA() {
        return a;
    }

    public ColourHolder setR(int r) {
        this.r = r;
        return this;
    }

    public ColourHolder setB(int b) {
        this.b = b;
        return this;
    }

    public ColourHolder setG(int g) {
        this.g = g;
        return this;
    }

    public ColourHolder setA(int a) {
        this.a = a;
        return this;
    }

    @Override
    public ColourHolder clone() {
        return new ColourHolder(r, g, b, a);
    }

    public Color toJavaColour() {
        return new Color(r, g, b, a);
    }
}

File: src/main/java/me/zeroeightsix/kami/util/colourUtils/ColourHolder.kt
package me.zeroeightsix.kami.util.colourUtils

import org.lwjgl.opengl.GL11
import java.awt.Color
import kotlin.math.max
import kotlin.math.min

/**
 * Created by Gebruiker on 18/04/2017.
 * Updated by Xiaro on 09/08/20
 */
class ColourHolder {
    var r = 0
    var g = 0
    var b = 0
    var a = 0

    constructor(r: Int, g: Int, b: Int) {
        this.r = r
        this.g = g
        this.b = b
        a = 255
    }

    constructor(r: Int, g: Int, b: Int, a: Int) {
        this.r = r
        this.g = g
        this.b = b
        this.a = a
    }

    constructor(color: Color) {
        this.r = color.red
        this.g = color.green
        this.b = color.blue
        this.a = color.alpha
    }

    fun brighter(): ColourHolder {
        return ColourHolder(min(r + 10, 255), min(g + 10, 255), min(b + 10, 255), a)
    }

    fun darker(): ColourHolder {
        return ColourHolder(max(r - 10, 0), max(g - 10, 0), max(b - 10, 0), a)
    }

    fun setGLColour() {
        setGLColour(-1, -1, -1, -1)
    }

    fun setGLColour(dr: Int, dg: Int, db: Int, da: Int) {
        GL11.glColor4f((if (dr == -1) r else dr).toFloat() / 255, (if (dg == -1) g else dg).toFloat() / 255, (if (db == -1) b else db).toFloat() / 255, (if (da == -1) a else da).toFloat() / 255)
    }

    fun becomeHex(hex: Int) {
        this.r = hex and 0xFF0000 shr 16
        this.g = hex and 0xFF00 shr 8
        this.b = hex and 0xFF
        this.a = 255
    }

    fun fromHex(hex: Int): ColourHolder {
        val n = ColourHolder(0, 0, 0)
        n.becomeHex(hex)
        return n
    }

    fun toHex(): Int {
        return 0xff shl 24 or (r and 0xff shl 16) or (g and 0xff shl 8) or (b and 0xff)
    }

    fun toJavaColour(): Color {
        return Color(r, g, b, a)
    }

    fun clone(): ColourHolder {
        return ColourHolder(r, g, b, a)
    }
}